FirebaseAppCheck/Sources/AppAttestProvider/FIRAppAttestProvider.m

```diff
@@ -138,23 +179,137 @@ - (void)getTokenWithCompletion:(void (^)(FIRAppCheckToken *_Nullable, NSError *_
 
                 return [self attestKey:keyID challenge:challenge];
               })
+      // TODO: Handle a possible key rejection - generate another key.
       .thenOn(self.queue,
               ^FBLPromise<FIRAppCheckToken *> *(FIRAppAttestKeyAttestationResult *result) {
                 // 5. Exchange the attestation to FAC token.
                 return [self.APIService appCheckTokenWithAttestation:result.attestation
                                                                keyID:result.keyID
                                                            challenge:result.challenge];
-              })
-      // 6. Call the handler with the result.
-      .then(^FBLPromise *(FIRAppCheckToken *token) {
-        handler(token, nil);
-        return nil;
-      })
-      .catch(^(NSError *error) {
-        handler(nil, error);
-      });
+              });
+}
+
+#pragma mark - Token refresh sequence
+
+- (FBLPromise<FIRAppCheckToken *> *)refreshTokenWithKeyID:(NSString *)keyID
+                                                 artifact:(NSData *)artifact {
+  // TODO: Implement (b/186438346).
+  return [FBLPromise resolvedWith:nil];
 }
 
+#pragma mark - State handling
+
+/// Calculates and returns current `FIRAppAttestAttestationState`.
+/// @return A promise that is resolved with FIRAppAttestProviderState with the state and associated
+/// data (e.g. key ID).
+- (FBLPromise<FIRAppAttestProviderState *> *)attestationState {
+  // Use a local variable to store App Attest key ID that may be fetched in the middle of the
+  // pipeline but may needed later. It simplifies chaining a bit.
+  __block NSString *appAttestKeyID;
+
+  return
+      // 1. Check if App Attest is supported.
+      [self isAppAttestSupported]
+          .recoverOn(self.queue,
+                     ^NSError *(NSError *error) {
+                       // App Attest is not supported.
+
+                       __auto_type state =
+                           [[FIRAppAttestProviderState alloc] initUnsupportedWithError:error];
+                       // Throw error to interrupt the pipeline earlier.
+                       return [self errorWithState:state];
+                     })
+
+          // 2. Check for stored key ID of the generated App Attest key pair.
+          .thenOn(self.queue,
+                  ^FBLPromise<NSString *> *(id result) {
+                    return [self appAttestKeyIDOrErrorWithState];
+                  })
```

Comment:
It took me awhile to understand this chain of operations but I think I now understand what is going on. I think it LGTM but I do have a suggestion that might improve clarity:

Make the each operation return a `FIRAppAttestProviderState` so it's like we a progressively building/transitioning towards a goal state. And use `catch` block at the end to catch any errors that occur along the way and return a corresponding state for that error.

Here's some pseudo objc to maybe illustrate what such a flow could look like. 
```objc
return [self isAppAttestSupported]                    // Promise resolved an initialState or error
    .thenOn(queue, ^Promise<State *> *(State *initialState)  {
        return [State nextStateFrom: initialState];   // Promise resolved with keyIDState or error.
    }
    .thenOn(queue, ^Promise<State *> *(State *keyIDState)  {
        return [State nextStateFrom: keyIDState];     // Promise resolved with artifactState or error.
    }
    .thenOn(queue, ^Promise<State *> *(State *artifactState)  {
        return [State nextStateFrom: artifactState];  // Promise resolved with finalState or error.
    }
    .thenOn(queue, ^Promise<State *> *(State * finalState)  {
        return [Promise resolvedWith: finalState];
    }
    .catch(error) {
        return [self stateFromError: error];
    }
``` 

This might require some additional refactoring so Im happy to help flush it out more if you want to explore it more.

Reply:
Thank you for your thorough review and the suggestion.

The main disadvantage with the described approach is that it will always do n number of steps and async dispatches even if we new the result at the very beginning. 

The only way to interrupt the promise chain/pipeline I found is to throw an error, which I tried to used. But yeah, it’s not ideal… I’ll take another look at what we can do here.  More suggestions are appreciated.

Reply:
we potentially may use ‘await’  promises API on another queue to make it in kind of sync way then early exit would be easier to express. I’ll try it out.

Reply:
I like the version with `FBLPromiseAwait()` (the current one) the most so far. WDYT?

Thank you for challenging readability here, it definitely is leading to improvements!